rclcpp/include/rclcpp/publisher_base.hpp

```diff
@@ -203,6 +204,73 @@ class PublisherBase : public std::enable_shared_from_this<PublisherBase>
   std::vector<rclcpp::NetworkFlowEndpoint>
   get_network_flow_endpoints() const;
 
+  /// Wait until all published message data is acknowledged or until the specified timeout elapses.
+  /**
+   * This function waits until all published message data were acknowledged by all subscribers or
+   * timeout.
+   *
+   * timeout must be less than std::chrono::nanoseconds::max().
+   * If the timeout is negative then this function will block indefinitely until all published
+   * message data were acknowledged.
+   * If the timeout is 0 then this function will be non-blocking; checking all published message
+   * data were acknowledged (If acknowledged, return true. Otherwise, return false),
+   * but not waiting.
+   * If the timeout is greater than 0 then this function will return after that period of time has
+   * elapsed (return false) or all published message data were acknowledged (return true).
+   *
+   * This function only waits for acknowledgments if the publisher's QOS profile is RELIABLE.
+   * Otherwise this function will immediately return `true`.
+   *
+   * \param[in] timeout the duration to wait for all published message data were acknowledged.
+   * \return `true` if all published message data were acknowledged before timeout, otherwise
+   *   `false`.
+   * \throws rclcpp::exceptions::RCLError if middleware doesn't support or internal error occurs
+   * \throws std::invalid_argument if timeout is greater than nanoseconds::max()
+   */
+  template<typename DurationRepT = int64_t, typename DurationT = std::milli>
+  bool
+  wait_for_all_acked(
+    std::chrono::duration<DurationRepT, DurationT> timeout =
+    std::chrono::duration<DurationRepT, DurationT>(-1)) const
+  {
+    // Casting to a double representation might lose precision and allow the check below to succeed
+    // but the actual cast to nanoseconds fail. Using 1 DurationT worth of nanoseconds less than max
+    constexpr auto maximum_safe_cast_ns =
+      std::chrono::nanoseconds::max() - std::chrono::duration<DurationRepT, DurationT>(1);
+
+    // If period is greater than nanoseconds::max(), the duration_cast to nanoseconds will overflow
+    // a signed integer, which is undefined behavior. Checking whether any std::chrono::duration is
+    // greater than nanoseconds::max() is a difficult general problem. This is a more conservative
+    // version of Howard Hinnant's (the <chrono> guy>) response here:
+    // https://stackoverflow.com/a/44637334/2089061
+    // However, this doesn't solve the issue for all possible duration types of period.
+    // Follow-up issue: https://github.com/ros2/rclcpp/issues/1177
+    constexpr auto ns_max_as_double =
+      std::chrono::duration_cast<std::chrono::duration<double, std::chrono::nanoseconds::period>>(
+      maximum_safe_cast_ns);
+    if (timeout > ns_max_as_double) {
+      throw std::invalid_argument{
+              "timeout must be less than std::chrono::nanoseconds::max()"};
+    }
```

Comment:
according to std::chrono::duration_cast() documentation in cppreference:

> Casting from a floating-point duration to an integer duration is subject to undefined behavior when the floating-point value is NaN, infinity, or too large to be representable by the target's integer type. Otherwise, casting to an integer duration is subject to truncation as with any static_cast to an integer type.

so I see the point of this, but the code it's a bit hard to follow and we need that in other places.
I would rather have a helper function in `rcpputils` "safe_duration_cast()`, and use that here.

For the moment, I would rather use std::chrono::duration_cast(), as we're already doing the same elsewhere.

Reply:
I also think the code can be simplified:

```suggestion
    if (timeout > std::chrono::nanoseconds::max()) {
      throw std::invalid_argument{
              "timeout must be less than std::chrono::nanoseconds::max()"};
    }
```

see https://en.cppreference.com/w/cpp/chrono/duration/operator_cmp.

Reply:
I think simplified codes may introduce some bugs.  
I will further confirm it.

Reply:
If timeout is  std::chrono::hours(10000000), it can pass (timeout > std::chrono::nanoseconds::max()).   
But we will get  a negative value after converting as below code  
int64_t v = (std::chrono::duration_cast<std::chrono::nanoseconds>(timeout)).count()  

So if use simplified way, it will include bugs.  
 I tend to adopt your below suggestion.  What do you think ? @ivanpauno 

> I would rather have a helper function in rcpputils "safe_duration_cast()`, and use that here.

Reply:
> If timeout is std::chrono::hours(10000000), it can pass (timeout > std::chrono::nanoseconds::max()).
But we will get a negative value after converting as below code

How?
`timeout > std::chrono::nanoseconds::max()` is true so we will throw an exception and the cast will never happen.

There's a problem if `timeout > std::chrono::nanoseconds::max()` is false and the cast fails anyway (because of rounding issues). I'm really not sure of that, cppreference doesn't document any guarantee.

> I tend to adopt your below suggestion. What do you think ? @ivanpauno

SGTM

Reply:
>> If timeout is std::chrono::hours(10000000), it can pass (timeout > std::chrono::nanoseconds::max()).
>> But we will get a negative value after converting as below code
>
> How?
> timeout > std::chrono::nanoseconds::max() is true so we will throw an exception and the cast will never happen.

Sorry. Make you confused. 
I means `timeout > std::chrono::nanoseconds::max()` is false. We get  a negative value after the cast is called.

Reply:
>> I tend to adopt your below suggestion. What do you think ? @ivanpauno
>
> SGTM

I created a PR ros2/rcpputils#145. Please help to review.